### PR TITLE
Implement kafka-streams-groups.sh --list

### PR DESCRIPTION
**Pull request by aliehsaeedii**
Implement the core of `kafka-streams-groups.sh`
Implement `--list` and its options: (only `--state`)

### CODE REVIEW

File: tools/src/main/java/org/apache/kafka/tools/streams/StreamsGroupCommand.java [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+import joptsimple.OptionException;
+
+public class StreamsGroupCommand {
+
+    public static void main(String[] args) {
+        StreamsGroupCommandOptions opts = new StreamsGroupCommandOptions(args);
+        try {
+            opts.checkArgs();
+
+            CommandLineUtils.maybePrintHelpOrVersion(opts, "This tool helps to list all streams groups.");
```
Comment (aliehsaeedii) [inline tools/src/main/java/org/apache/kafka/tools/streams/StreamsGroupCommand.java:51]: The message should change as new command are added.

File: tools/src/main/java/org/apache/kafka/tools/streams/StreamsGroupCommand.java [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+
+            // should have exactly one action
+            long actions = Stream.of(opts.listOpt).filter(opts.options::has).count();
+            if (actions != 1)
+                CommandLineUtils.printUsageAndExit(opts.parser, "Command must include exactly one action: --list.");
```
Comment (aliehsaeedii) [inline tools/src/main/java/org/apache/kafka/tools/streams/StreamsGroupCommand.java:56]: The same here.

File: tools/src/main/java/org/apache/kafka/tools/streams/StreamsGroupCommandOptions.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+import org.apache.kafka.server.util.CommandDefaultOptions;
+import org.apache.kafka.server.util.CommandLineUtils;
+
+import joptsimple.OptionSpec;
+
+
+public class StreamsGroupCommandOptions extends CommandDefaultOptions {
+    public static final String BOOTSTRAP_SERVER_DOC = "REQUIRED: The server(s) to connect to.";
+    public static final String LIST_DOC = "List all share groups.";
```
Comment (lucasbru) [inline tools/src/main/java/org/apache/kafka/tools/streams/StreamsGroupCommandOptions.java]: "List all streams groups"

File: tools/src/main/java/org/apache/kafka/tools/streams/StreamsGroupCommandOptions.java [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+ * limitations under the License.
+ */
+package org.apache.kafka.tools.streams;
+
+import org.apache.kafka.server.util.CommandDefaultOptions;
+import org.apache.kafka.server.util.CommandLineUtils;
+
+import joptsimple.OptionSpec;
+
+
```
Comment (lucasbru) [inline tools/src/main/java/org/apache/kafka/tools/streams/StreamsGroupCommandOptions.java]: double new line

File: tools/src/main/java/org/apache/kafka/tools/streams/StreamsGroupCommandOptions.java [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+public class StreamsGroupCommandOptions extends CommandDefaultOptions {
+    public static final String BOOTSTRAP_SERVER_DOC = "REQUIRED: The server(s) to connect to.";
+    public static final String LIST_DOC = "List all share groups.";
+    public static final String NL = System.lineSeparator();
+    public static final String TIMEOUT_MS_DOC = "The timeout that can be set for some use cases. For example, it can be used when describing the group " +
+        "to specify the maximum amount of time in milliseconds to wait before the group stabilizes.";
+    public static final String COMMAND_CONFIG_DOC = "Property file containing configs to be passed to Admin Client.";
+    public static final String DRY_RUN_DOC = "Only show results without executing changes on share groups. Supported operations: reset-offsets.";
```
Comment (lucasbru) [inline tools/src/main/java/org/apache/kafka/tools/streams/StreamsGroupCommandOptions.java]: We do not need --dry-run yet, as it's only supported by reset-offsets. Also, this shouldn't mention share groups.

File: tools/src/main/java/org/apache/kafka/tools/streams/StreamsGroupCommandOptions.java [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+    public static final String EXECUTE_DOC = "Execute operation. Supported operations: reset-offsets.";
```
Comment (lucasbru) [inline tools/src/main/java/org/apache/kafka/tools/streams/StreamsGroupCommandOptions.java]: remove

File: tools/src/main/java/org/apache/kafka/tools/streams/StreamsGroupCommandOptions.java [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+    public static final String STATE_DOC = "When specified with '--describe', includes the state of the group." + NL +
```
Comment (lucasbru) [inline tools/src/main/java/org/apache/kafka/tools/streams/StreamsGroupCommandOptions.java]: Seems like this shouldn't mention `--describe` yet.

File: tools/src/main/java/org/apache/kafka/tools/streams/StreamsGroupCommandOptions.java [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+        "When specified with '--list', it displays the state of all groups. It can also be used to list groups with specific states. " +
+        "Valid values are Empty, Stable and Dead.";
```
Comment (lucasbru) [inline tools/src/main/java/org/apache/kafka/tools/streams/StreamsGroupCommandOptions.java]: There are more possible states than just `Empty, Stable and Dead`.

File: tools/src/main/java/org/apache/kafka/tools/streams/StreamsGroupCommandOptions.java [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+        stateOpt = parser.accepts("state", STATE_DOC)
+            .availableIf(listOpt)
+            .withOptionalArg()
+            .ofType(String.class);
+
+        options = parser.parse(args);
+    }
+
+    public void checkArgs() {
+        CommandLineUtils.maybePrintHelpOrVersion(this, "This tool helps to list, describe, reset and delete share groups.");
```
Comment (lucasbru) [inline tools/src/main/java/org/apache/kafka/tools/streams/StreamsGroupCommandOptions.java]: Mentions share groups.
reset and delete and describe are not probided yet

File: tools/src/test/java/org/apache/kafka/tools/streams/StreamsGroupCommandTest.java [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+
+import joptsimple.OptionException;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+public class StreamsGroupCommandTest {
```
Comment (lucasbru) [inline tools/src/test/java/org/apache/kafka/tools/streams/StreamsGroupCommandTest.java]: We'll also want to have an integration test for the tool as well, similar to `ListConsumerGroupTest`, which actually runs a cluster. Maybe we could rename this to `StreamsGroupCommandUnitTest` and add a `StreamsGroupCommandTest` that uses a similar pattern as `ListConsumerGroupTest`?

**Review by lucasbru — APPROVED** [review]
Thanks for the PR! I left some comments

**Review by lucasbru — COMMENTED** [review]
Thanks for the PR! I left some comments

**Comment by aliehsaeedii** [comment]
> Thanks for the PR! I left some comments

Thanks, @lucasbru. I addressed the nits. I think we can merge this PR and meanwhile I work on the integration test.

File: tools/src/test/java/org/apache/kafka/tools/streams/StreamsGroupCommandUnitTest.java [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+        assertEquals(new HashSet<>(Collections.singletonList(GroupState.STABLE)), result);
+
+        result = StreamsGroupCommand.groupStatesFromString("DEAD");
+        assertEquals(new HashSet<>(Collections.singletonList(GroupState.DEAD)), result);
+        result = StreamsGroupCommand.groupStatesFromString("dead");
+        assertEquals(new HashSet<>(Collections.singletonList(GroupState.DEAD)), result);
+
+
+        result = StreamsGroupCommand.groupStatesFromString("PREPARINGREBALANCE");
+        assertEquals(new HashSet<>(Collections.singletonList(GroupState.PREPARING_REBALANCE)), result);
```
Comment (lucasbru) [inline tools/src/test/java/org/apache/kafka/tools/streams/StreamsGroupCommandUnitTest.java]: @aliehsaeedii as per discussion, can you please remove preparing_rebalance and completing rebalance from the possible states in `GroupState` for streams, and this test? Thanks!

**Review by lucasbru — APPROVED** [review]
Thanks for the PR! LGTM

**Pull request merged**